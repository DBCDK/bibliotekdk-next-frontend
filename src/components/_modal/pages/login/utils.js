import { LOGIN_MODE } from "@/components/_modal/pages/loanerform/LoanerForm";
import Translate from "@/components/base/translate";

/**
 * If we order a physical copy or a digital copy, we always show the loanerform
 * If the mode neither of the two modes above, we check if the library supports borrowerCheck
 * If not, we show the loginNotSupported modal, we can order stuff for these libraries, but we cannot login/validate via them
 * @param {string} mode
 * @returns {boolean}
 */
export const showLogin = (mode) => {
  return mode === LOGIN_MODE.ORDER_PHYSICAL && mode === LOGIN_MODE.SUBSCRIPTION;
};

export const isOrderPossible = ({ mode, branch }) => {
  if (!branch) return false;
  // QUICKFIX - .. to avoid api check ..  all public libraries have access to dda - no other
  // Færøerne skal ikke tjekkes
  if (mode === LOGIN_MODE.DDA) {
    return (
      branch?.agencyId?.startsWith("7") && branch.orderPolicy?.orderPossible
    );
  }
  return branch.orderPolicy?.orderPossible;
};

/**
 * Open login modal
 * @param {obj} modal
 * @param {string} title
 * @param {string} mode
 */
export function openLoginModal({
  modal,
  title = Translate({
    context: "header",
    label: "login",
  }),
  mode = LOGIN_MODE.PLAIN_LOGIN,
}) {
  console.log("utils", mode);
  modal.push("login", {
    title: title,
    mode: mode,
  });
}
