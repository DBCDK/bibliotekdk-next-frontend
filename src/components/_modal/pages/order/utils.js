import styles from "@/components/_modal/pages/order/Order.module.css";

/**
 *
 * @param {*} value
 * @param {*} valid
 * @param updateLoanerInfo
 * @param setMail
 */
export function onMailChange(value, valid, updateLoanerInfo, setMail) {
  valid.status &&
    updateLoanerInfo &&
    updateLoanerInfo({ userParameters: { userMail: value } });
  // update mail in state
  setMail({ value, valid });
}

/**
 *
 * @param pid
 * @param work
 * @param singleManifestation
 * @returns {*[]}
 */
export function getOrderPids(pid, work, singleManifestation) {
  // Work props
  const manifestations = work?.manifestations?.all;

  // Material by pid
  const material = manifestations?.filter(
    (manifestation) => manifestation.pid === pid
  );

  const materialsSameType = manifestations?.filter((manifestation) => {
    return (
      manifestation?.materialType === material?.materialType &&
      (!material?.accessTypes?.find(
        (accessType) => accessType?.display !== "fysisk"
      ) ||
        material?.access?.find((singleAccess) => singleAccess.loanIsPossible))
    );
  });

  let orderPids;
  if (singleManifestation) {
    orderPids = [pid];
  } else {
    orderPids = materialsSameType?.map((m) => m.pid);
  }
  return orderPids;
}

export function extractClassNameAndMessage(validated, failedSubmission) {
  // Get email messages (from validate object)
  const emailStatus = validated?.details?.hasMail?.status;
  const errorMessage = validated?.details?.hasMail?.message;

  // Check for email validation and email error messages
  const hasEmail = !!validated?.details?.hasMail?.status;

  // Email validation class
  const validClass =
    failedSubmission && !emailStatus ? styles.invalid : styles.valid;
  const customInvalidClass =
    failedSubmission && !emailStatus ? styles.invalidInput : "";

  // Set email input message if any
  const message = failedSubmission && errorMessage;

  const actionMessage =
    failedSubmission &&
    (validated?.details?.requireYear?.message ||
      (!hasEmail && validated?.details?.hasMail?.message));

  const invalidClass = actionMessage ? styles.invalid : "";

  return {
    emailStatus,
    errorMessage,
    hasEmail,
    validClass,
    customInvalidClass,
    message,
    actionMessage,
    invalidClass,
  };
}
