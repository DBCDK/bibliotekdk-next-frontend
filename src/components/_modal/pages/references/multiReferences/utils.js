/**
 * Splits the list materialsMissingEdition into two lists:
 * - newPeriodicaFiltered: list of periodica materials, for which we cannot create references
 * - listWithoutPeriodica: list of remaining materials, for which we can create references
 * @param {Array<Object>} materialsMissingEdition
 * @returns {Object}
 */
export const splitList = (materialsMissingEdition) => {
  const periodicaManifestations = [];
  const nonPeriodicaManifestations = materialsMissingEdition.map((item) => {
    const manifestations = item.manifestations;
    // If 1 option, select it
    if (manifestations.length === 1) {
      const singleManifestation = manifestations[0];
      if (singleManifestation.workTypes?.[0] === "PERIODICA") {
        //TODO What about journals??
        // Periodica -
        periodicaManifestations.push(singleManifestation);
        return {
          ...item,
          toFilter: true,
        };
      }
      return {
        ...item,
        chosenPid: singleManifestation.pid,
      };
    } else return item;
  });
  return {
    periodicaManifestations,
    nonPeriodicaManifestations,
  };
};
