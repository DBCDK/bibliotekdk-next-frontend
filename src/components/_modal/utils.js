// elements we want
const elements = [
  "a",
  "input:not([type='hidden'])",
  "button",
  "textarea",
  "[tabindex]",
];

// currently not in use
export function tabVisibility(container, isVisible) {
  // build query with elements
  const query = elements.join(", ");

  // Find mathing elements according to elements and select string
  const matchedElements = Object.values(container.querySelectorAll(query));

  matchedElements.forEach((el) => {
    const tabIndex = el.getAttribute("tabindex") || false;
    const savedTabIndex = el.getAttribute("data-tabindex") || false;

    el.setAttribute(
      "tabindex",
      isVisible ? savedTabIndex || tabIndex || "0" : "-1"
    );

    if (!savedTabIndex) {
      el.setAttribute("data-tabindex", tabIndex);
    }
  });
}

/**
 * Function to trap Tab inside modal
 *
 * @param {obj} event current target element (focused element)
 * @param {obj} container container to trap Tab in (modal)
 *
 * https://medium.com/@islam.sayed8/trap-focus-inside-a-modal-aa5230326c1b
 * https://medium.com/@seif_ghezala/how-to-create-an-accessible-react-modal-5b87e6a27503
 */

export function handleTab(event, container) {
  event.preventDefault();

  const el = container.querySelector("div.modal_page.page-current");

  // Custom select string to tail after each element in list
  // Dont select elements with display:none inline style
  // Dont select hidden elements
  // Dont select elements with tabindex -1
  const select =
    ":not([aria-hidden='true']):not([tabindex='-1']):not([style*='display:none']):not([style*='display: none']):not([disabled])";

  // build query with elements
  const query = elements.join(select + ", ") + select;

  // Find mathing elements according to elements and select string
  const matchedElements = Object.values(el.querySelectorAll(query));

  // Remove elements which parent is not visible
  const visibleChildren = matchedElements.filter(
    (el) => el.offsetParent !== null
  );

  // Remove all elements which is display:none in computedStyles (css stylesheet)
  // OBS! Performance Warning! getComputedStyle func can be slow
  const sequence = visibleChildren.filter(
    (el) => getComputedStyle(el).display !== "none"
  );

  // Debug -> remove me in future
  console.log("Debug", { sequence });

  if (sequence.length < 1) {
    return;
  }

  const backward = event.shiftKey;
  const first = sequence[0];
  const last = sequence[sequence.length - 1];

  // wrap around first to last, last to first
  const source = backward ? first : last;
  const target = backward ? last : first;

  if (source === event.target) {
    target.focus();
    return;
  }

  // find current position in tabsequence
  let currentIndex;
  const found = sequence.some(function (element, index) {
    if (element !== event.target) {
      return false;
    }

    currentIndex = index;
    return true;
  });

  if (!found) {
    // redirect to first as we're not in our tabsequence
    first.focus();
    return;
  }

  // shift focus to previous/next element in the sequence
  const offset = backward ? -1 : 1;
  sequence[currentIndex + offset].focus();
}

/**
 * Function to get random color by random string
 *
 *  @param {string} id
 *  @param {array} colors
 */
export function toColor(
  id,
  colors = ["#ffe7e0", "#e5c7bd", "#f4efdd", "#eed9b0", "#b7dee2"]
) {
  let hash = 0;
  if (id.length === 0) {
    return hash;
  }
  for (let i = 0; i < id.length; i++) {
    hash = id.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}

/**
 * Function to get scrollY (scroll distance from top)
 *
 * @returns {int}
 */
function getScrollYPos() {
  // Get scrollY (all browsers)
  var doc = document.documentElement;
  var top = (window.pageYOffset || doc.scrollTop) - (doc.clientTop || 0);
  return top;
}

/**
 * Function to handle scrollLock on body
 *
 * @param {bool} shouldLockScroll
 *
 */
let scrollY = 0;
export function scrollLock(shouldLockScroll) {
  const body = document.body;
  const layout = document.getElementById("layout");

  // We need booth
  if (!body || !layout) {
    return;
  }

  const isLocked = body.classList.contains("lockScroll");

  // Add "lock" class and add "fake" scrollY position to body
  if (shouldLockScroll && !isLocked) {
    scrollY = getScrollYPos();
    layout.style.marginTop = `-${scrollY}px`;
    body.classList.add("lockScroll");
  }
  // Remove "lock", remove "fake" scrollY position
  // + Scroll back to the scrollY position - same as before the modal was triggered
  else if (!shouldLockScroll && isLocked) {
    body.classList.remove("lockScroll");
    layout.style.marginTop = `auto`;
    window.scrollTo(0, scrollY);
  }
}
