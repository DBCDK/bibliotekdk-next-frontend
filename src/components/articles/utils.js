import { orderBy } from "lodash";

/**
 * The Article page React component
 *
 * @param {obj} props
 * See propTypes for specific props and types
 *
 * @returns {component}
 */
export function sortArticles(articles) {
  // latest articles first
  return orderBy(articles, ["entityCreated"], ["desc"]);
}
