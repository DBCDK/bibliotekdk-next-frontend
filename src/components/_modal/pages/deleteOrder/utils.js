import * as orderMutations from "@/lib/api/order.mutations";
import Translate from "@/components/base/translate";

export async function handleCancelOrder(orderId, agencyId, orderMutation) {
  try {
    return orderMutation.post(
      orderMutations.cancelOrder({
        orderId,
        agencyId,
      })
    );
  } catch (e) {
    console.log("could not delete orderq", e);

    return {
      error: `Could not delete order with orderId ${orderId} and agencyId ${agencyId} due to error ${e}`,
    };
  }
}

/**
 * Opens delete order modal
 * @param {*} param0.modal
 * @param {boolean} param0.mobile
 * @param {Date} param0.pickUpExpiryDate
 * @param {string} param0.id
 * @param {string} param0.agencyId
 * @param {string} param0.orderMutation
 * @returns success status, error message
 */
export function onClickDelete({
  modal,
  mobile,
  pickUpExpiryDate,
  id,
  agencyId,
  orderMutation,
  onCloseModal,
}) {
  modal.push("deleteOrder", {
    label: Translate({ context: "profile", label: "delete-order" }),
    mobile: mobile,
    isReadyToPickup: !!pickUpExpiryDate,
    orderId: id,
    agencyId: agencyId,
    orderMutation: orderMutation,
    onClose: onCloseModal,
  });
}
