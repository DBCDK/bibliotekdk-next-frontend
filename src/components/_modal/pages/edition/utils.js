import { getIsPeriodicaLike } from "@/lib/utils";

/**
 * Get a coverimage to use from given manifestations - from moreinfo OR default cover service
 * @param manifestations
 * @returns {{detail}|{detail: null}}
 */
export function editionCover(manifestations) {
  const manifestationWithCover = manifestations?.find(
    (manifestation) => manifestation?.cover.detail
  );
  return manifestationWithCover
    ? { detail: manifestationWithCover.cover.detail }
    : { detail: null };
}

export function inferAccessTypes(
  work,
  periodicaForm,
  initialPickupBranch,
  manifestationsBeforeCheck = null
) {
  const manifestations =
    manifestationsBeforeCheck || work?.manifestations?.all || [];

  const isArticle = !!work?.workTypes?.find(
    (workType) => workType.toLowerCase() === "article"
  );

  const isPeriodicaLike = getIsPeriodicaLike(
    work?.workTypes,
    work?.materialTypes
  );
  const isArticleRequest =
    !!periodicaForm?.titleOfComponent ||
    !!periodicaForm?.authorOfComponent ||
    !!periodicaForm?.pagination;

  const isDigitalCopy = !!manifestations?.find((m) => {
    return !!m?.access?.find((a) => a.issn);
  });

  const availableAsDigitalCopy =
    isDigitalCopy &&
    initialPickupBranch?.digitalCopyAccess &&
    (isPeriodicaLike ? isArticleRequest : true);

  const isPhysical = !!manifestations?.find((m) => {
    return m?.accessTypes?.find(
      (accessType) => accessType?.code === "PHYSICAL"
    );
  });
  const availableAsPhysicalCopy =
    isPhysical &&
    initialPickupBranch?.pickupAllowed &&
    initialPickupBranch?.orderPolicy?.orderPossible;

  const requireDigitalAccess = isDigitalCopy && !isPhysical;

  return {
    isArticle,
    isPeriodicaLike,
    isArticleRequest,
    isDigitalCopy,
    availableAsDigitalCopy,
    isPhysical,
    availableAsPhysicalCopy,
    requireDigitalAccess,
  };
}
