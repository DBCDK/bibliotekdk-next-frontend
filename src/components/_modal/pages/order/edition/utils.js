import { getIsPeriodicaLike } from "@/lib/utils";

export function inferCheckers(
  work,
  context,
  manifestations,
  initialPickupBranch
) {
  const isArticle = work?.workTypes?.find(
    (workType) => workType.toLowerCase() === "article"
  );
  const isPeriodicaLike = getIsPeriodicaLike(
    work?.workTypes,
    work?.materialTypes
  );
  const isArticleRequest =
    !!context?.periodicaForm?.titleOfComponent ||
    !!context?.periodicaForm?.authorOfComponent ||
    !!context?.periodicaForm?.pagination;
  const isDigitalCopy = !!manifestations?.find((m) =>
    m?.access?.find((entry) => entry.issn)
  );
  const availableAsDigitalCopy =
    initialPickupBranch?.pickupBranch?.digitalCopyAccess &&
    isDigitalCopy &&
    (isPeriodicaLike ? isArticleRequest : true);

  return {
    isArticle,
    isPeriodicaLike,
    isArticleRequest,
    isDigitalCopy,
    availableAsDigitalCopy,
  };
}
