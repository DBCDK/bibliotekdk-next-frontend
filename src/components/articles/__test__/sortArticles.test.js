import { sortArticles } from "../utils";
import data from "./mock/articles.mock";

test("sortArticles desc", () => {
  const actual = sortArticles(data);
  console.log("actual", actual);
  expect(actual).toMatchSnapshot();
});
