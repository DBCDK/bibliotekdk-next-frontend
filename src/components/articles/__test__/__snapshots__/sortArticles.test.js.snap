// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortArticles desc 1`] = `
Array [
  Object {
    "entityCreated": "2021-02-25T14:40:40+0100",
    "fieldImage": Object {
      "alt": "some image alt",
      "title": "some image title",
      "url": "/img/bibdk-hero-scaled.jpeg",
    },
    "fieldRubrik": "På bibliotek.dk søger du i alle landets fysiske og digitale biblioteker. Det du ønsker kan du nemt få leveret til dit lokale bibliotek eller tilgå direkte online.",
    "nid": 4,
    "title": "Nyeste artikel",
  },
  Object {
    "entityCreated": "2021-01-20T14:38:46+0100",
    "fieldImage": Object {
      "alt": "some image alt",
      "title": "some image title",
      "url": "/img/bibdk-hero-scaled.jpeg",
    },
    "fieldRubrik": "På bibliotek.dk søger du i alle landets fysiske og digitale biblioteker. Det du ønsker kan du nemt få leveret til dit lokale bibliotek eller tilgå direkte online.",
    "nid": 3,
    "title": "Ny artikel",
  },
  Object {
    "entityCreated": "2021-01-13T13:47:27+0100",
    "fieldImage": Object {
      "alt": "some image alt",
      "title": "some image title",
      "url": "/img/bibdk-hero-scaled.jpeg",
    },
    "fieldRubrik": "Online bibliotekarhjælp. Få råd og hjælp til alt fra informationssøgning til reservationer.",
    "nid": 2,
    "title": "Gammel artikel",
  },
  Object {
    "entityCreated": "2021-01-05T10:19:13+0100",
    "fieldImage": Object {
      "alt": "some image alt",
      "title": "some image title",
      "url": "/img/bibdk-hero-scaled.jpeg",
    },
    "fieldRubrik": "Læs mere om forfattere, musik og temaer. Se film, læs artikler, e- og lydbøger, og meget mere.",
    "nid": 1,
    "title": "Ældste artikel",
  },
]
`;
